demo/src/ddnet/writer.rs

```diff
@@ -0,0 +1,205 @@
+use crate::format;
+use gamenet_ddnet::msg;
+use gamenet_ddnet::snap_obj;
+use snapshot::snap;
+use std::collections::hash_map;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::convert::TryInto;
+use std::io;
+use std::mem;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum WriteError {
+    #[error(transparent)]
+    Inner(crate::WriteError),
+    #[error("Snap creation - {0:?}")]
+    SnapBuilder(snap::BuilderError),
+    #[error("Tick decreased or is negative")]
+    TooLowTickNumber,
+    #[error("Snap data does not fit into buffer")]
+    TooLargeSnap,
+    #[error("Net message data does not fit into buffer")]
+    TooLongNetMsg,
+}
+
+impl From<crate::WriteError> for WriteError {
+    fn from(value: crate::WriteError) -> Self {
+        Self::Inner(value)
+    }
+}
+
+impl From<snap::BuilderError> for WriteError {
+    fn from(value: snap::BuilderError) -> Self {
+        Self::SnapBuilder(value)
+    }
+}
+
+/// DDNet demo writer, for use with `gamenet_ddnet`.
+/// Automatically writes snapshot deltas.
+pub struct DemoWriter {
+    inner: crate::Writer,
+    /// To verify the monotonic increase
+    last_tick: i32,
+    /// For delta timing
+    last_keyframe: Option<i32>,
+    uuid_index: UuidIndex,
+    snap: snapshot::Snap,
+    builder: snap::Builder,
+    delta: snapshot::Delta,
+    buf: arrayvec::ArrayVec<[u8; format::MAX_SNAPSHOT_SIZE]>,
+    i32_buf: Vec<i32>,
+}
+
+struct UuidIndex {
+    index: HashMap<[u8; 16], u16>,
+    next_type_id: u16,
+    advertised: HashSet<[u8; 16]>,
+}
+
+impl UuidIndex {
+    fn new() -> Self {
+        Self {
+            index: HashMap::new(),
+            next_type_id: 32767,
```

Comment:
Where does this number come from?

Reply:
Resolved with UuidIndex rewrite and comment